File: BasicRxJavaSample/app/src/main/java/com/kinglloy/example/android/observability/ui/ViewModelFactory.java
package com.kinglloy.example.android.observability.ui;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.kinglloy.example.android.observability.UserDataSource;

/**
 * Factory for ViewModels
 *
 * @author jinyalin
 * @since 2018/1/21.
 */

public class ViewModelFactory implements ViewModelProvider.Factory {
    private final UserDataSource mDataSource;

    public ViewModelFactory(UserDataSource dataSource) {
        mDataSource = dataSource;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(UserViewModel.class)) {
            return (T) new UserViewModel(mDataSource);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: BasicRxJavaSampleKotlin/app/src/main/java/com/kinglloy/example/android/kotlin/ui/ViewModelFactory.kt
package com.kinglloy.example.android.kotlin.ui

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import com.kinglloy.example.android.kotlin.persistence.UserDao

/**
 * Factory for ViewModels
 * @author jinyalin
 * @since 2018/1/22.
 */
class ViewModelFactory(private val dataSource: UserDao) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(UserViewModel::class.java)) {
            return UserViewModel(dataSource) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }

}